kevlar/include/kevlar_bits/model/binomial/berry_inla.hpp

```diff
@@ -0,0 +1,368 @@
+
+            // calculate exceedance probabilities, integrated over sigma2
+            for (int i = 0; i < ARMS; i++) {
+                double normalized =
+                    (efficacy_thresholds[i] - t[i]) / theta_sigma[i];
+                double exc_sigma2 = 0.5 * (erf(-normalized * M_SQRT1_2) + 1);
+                exceedances[i] += exc_sigma2 * sigma2_post * quad_wts[s];
+            }
+        }
+
+        // normalize the sigma2_post distribution
+        double inv_sigma2_integral = 1.0 / sigma2_integral;
+        for (int i = 0; i < ARMS; i++) {
+            exceedances[i] *= inv_sigma2_integral;
+        }
+        return exceedances;
+    }
+};
+
+template <class ValueType, int ARMS>
+template <class _GenType, class _ValueType, class _UIntType,
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState
+    : SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                  _GridRangeType> {
+    struct SimState;
+
+    using base_t = SimGlobalStateFixedNDefault<_GenType, _ValueType, _UIntType,
+                                               _GridRangeType>;
+    using typename base_t::gen_t;
+    using typename base_t::grid_range_t;
+    using typename base_t::interface_t;
+    using typename base_t::uint_t;
+    using typename base_t::value_t;
+
+    using sim_state_t = SimState;
+
+   private:
+    using model_t = BerryINLA;
+    const model_t& model;
+
+   public:
+    SimGlobalState(const model_t& model, const grid_range_t& grid_range)
+        : base_t(model.n_arm_samples, grid_range), model(model) {}
+
+    std::unique_ptr<typename interface_t::sim_state_t> make_sim_state(
+        size_t seed) const override {
+        return std::make_unique<sim_state_t>(*this, seed);
+    }
+};
+
+template <class ValueType, int ARMS>
+template <class _GenType, class _ValueType, class _UIntType,
+          class _GridRangeType>
+struct BerryINLA<ValueType, ARMS>::SimGlobalState<
+    _GenType, _ValueType, _UIntType, _GridRangeType>::SimState
+    : base_t::sim_state_t {
+   private:
+    using outer_t = SimGlobalState;
+
+   public:
+    using base_t = typename outer_t::base_t::sim_state_t;
+    using typename base_t::interface_t;
+
+   private:
+    const outer_t& outer_;
+
+   public:
+    SimState(const outer_t& sgs, size_t seed)
+        : base_t(sgs, seed), outer_(sgs) {}
+
+    void simulate(Eigen::Ref<colvec_type<uint_t>> rej_len) override {
+        base_t::generate_data();
+        base_t::generate_sufficient_stats();
+
+        const auto& bits = outer_.bits();
+        const auto& gr = outer_.grid_range();
+
+        assert(gr.n_params() == ARMS);
+
+        const auto n_arm_samples = outer_.model.n_arm_samples;
+        const auto& critical_values = outer_.model.critical_values();
+
+        size_t pos = 0;
+        for (size_t grid_i = 0; grid_i < gr.n_gridpts(); ++grid_i) {
+            auto bits_i = bits.col(grid_i);
+
+            vec_t y(ARMS);
```

Comment:
For performance reasons, it would be ideal to replace any `vec_t` usage with a statically-sized vector (i.e. pass the template argument) and save the allocation.

No need to unless this is performance critical though.